azure/scope/managedmachinepool.go

```diff
@@ -154,6 +154,11 @@ func buildAgentPoolSpec(managedControlPlane *infrav1.AzureManagedControlPlane,
 		replicas = *machinePool.Spec.Replicas
 	}
 
+	subnetName := managedControlPlane.Spec.VirtualNetwork.Subnet.Name
+	if managedMachinePool.Spec.SubnetName != "" {
+		subnetName = managedMachinePool.Spec.SubnetName
```

Comment:
have you considered defaulting the subnetName on the managedMachinePool object too so it's persisted in the API and visible by the user?

Reply:
ack

Reply:
@CecileRobertMichon Thoughts about the above and let me know if i have to make the above changes.

Reply:
+1 to defaulting

Reply:
@LochanRn have you made the defaulting changes? If so, I'll go ahead and review the PR again!

Reply:
@CecileRobertMichon I tried to implement the above, I was facing intermittent issues while testing. The issue is because after the above implementation AzureManagedMachinePool has a couple of dependencies.

- before AzureManagedMachinePool is created AzureManagedControlPlane has to be created.
- capi-controller should set label _cluster.x-k8s.io/cluster-name_ on AzureManagedMachinePool and AzureManagedControlPlane.

This is because we need to get the AzureManagedControlPlane object to get AzureManagedControlPlane.Spec.VirtualNetwork.SubnetName. If AzureManagedControlPlane is not present AzureManagedMachinePool will be created but AzureManagedMachinePool.Spec.SubnetName value will not get set until the next update.

If the label is not present we cannot get AzureManagedControlPlane.

Reply:
The main issue is with the timing. When we create a cluster we apply all the manifests at once so at times the amcp used to get created first and even capi controller was also fast enough to set the clusterlabel before the machinepool was even created. Hence the ammp defaulting webhook was able to kick in at the right time and update the subnet value but for some cases it was not the same. ammp used to get created and it used to be stuck until the next update.  As the defaulting webhook would kick in only during creation or update.

Reply:
The implementation with which i tested was

```// Default implements webhook.Defaulter so a webhook will be registered for the type.
func (m *AzureManagedMachinePool) Default(cli client.Client) {
	if m.Labels == nil {
		m.Labels = make(map[string]string)
	}
	m.Labels[LabelAgentPoolMode] = m.Spec.Mode

	if m.Spec.Name == nil || *m.Spec.Name == "" {
		m.Spec.Name = &m.Name
	}

	if m.Spec.OSType == nil {
		m.Spec.OSType = to.StringPtr(DefaultOSType)
	}
	m.defaultSubnetName(cli)
}

func (m *AzureManagedMachinePool) defaultSubnetName(cli client.Client) {

	clusterName, ok := m.Labels[clusterv1.ClusterLabelName]
	if !ok {
		log.Errorf("capi has not yet set label %s",clusterv1.ClusterLabelName)
		return nil
	}

	// Fetch the amcp.
	amcp := &AzureManagedControlPlane{}
	key := client.ObjectKey{
		Namespace: m.Namespace,
		Name:      clusterName,
	}

	ctx := context.Background()

	if err := cli.Get(ctx, key, amcp); err != nil {
		log.Error(err)
		return nil
	}

	if amcp.Spec.VirtualNetwork.Name != "" {
		if m.Spec.SubnetName == "" {
			if amcp.Spec.VirtualNetwork.Subnet.Name == "" {
				log.Error("a virtualnetwork is defined but a subnet was undefined")
				return nil
			}
			m.Spec.SubnetName = amcp.Spec.VirtualNetwork.Subnet.Name
		}
	}
	return nil
}```

Reply:
okay !!